debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordState.java

```diff
@@ -143,6 +164,55 @@ public R apply(final R record) {
         }
     }
 
+    private R addSourceFields(String[] addSourceFields, R originalRecord, R unwrappedRecord) {
+        // Return if no source fields to add
+        if(addSourceFields == null) {
+            return unwrappedRecord;
+        }
+        
+        final Struct value = requireStruct(unwrappedRecord.value(), PURPOSE);
+        Struct source = ((Struct) originalRecord.value()).getStruct("source");
+        
+        // Get the updated schema from the cache, or create and cache if it doesn't exist
+        Schema updatedSchema = schemaUpdateCache.get(value.schema());
+        if (updatedSchema == null) {
+            updatedSchema = makeUpdatedSchema(value.schema(), addSourceFields);
+            schemaUpdateCache.put(value.schema(), updatedSchema);
+        }   
+
+        // Create the updated struct
+        final Struct updatedValue = new Struct(updatedSchema);
+        for (org.apache.kafka.connect.data.Field field : value.schema().fields()) {
+            updatedValue.put(field.name(), value.get(field));
+        }
+        for(String sourceField : addSourceFields) {
+            String fieldValue = source.schema().field(sourceField) == null ? "" : source.get(sourceField).toString();
+            updatedValue.put("__" + sourceField, fieldValue);
+        }
+
+        return unwrappedRecord.newRecord(
+            unwrappedRecord.topic(), 
+            unwrappedRecord.kafkaPartition(), 
+            unwrappedRecord.keySchema(), 
+            unwrappedRecord.key(), 
+            updatedSchema, 
+            updatedValue, 
+            unwrappedRecord.timestamp());
+    }
+
+    private Schema makeUpdatedSchema(Schema schema, String[] addSourceFields) {
+        final SchemaBuilder builder = SchemaUtil.copySchemaBasics(schema, SchemaBuilder.struct());
+        // Get fields from original schema
+        for (org.apache.kafka.connect.data.Field field : schema.fields()) {
+            builder.field(field.name(), field.schema());
+        }
+        // Add the requested source fields
+        for(String sourceField: addSourceFields) {
+            builder.field("__" + sourceField, SchemaBuilder.string());
```

Comment:
Instead of using string as the type, wouldn't it be better to use the original type of the source field?